### PR TITLE
DOC: add broadcasting article from scipy old-wiki

### LINKED ISSUE #12223: integrate content/images on broadcasting in docs
https://scipy.github.io/old-wiki/pages/EricsBroadcastingDoc explains broadcasting well (including useful diagrams) and is linked to from https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html. It is the only link to https://scipy.github.io/old-wiki left.

This content should be integrated in the user guide. There's also useful code to draw such diagrams at https://jakevdp.github.io/PythonDataScienceHandbook/02.05-computation-on-arrays-broadcasting.html

### PR DESCRIPTION
**Pull request by mattip**
Closes #12223, see rational there.

### CODE REVIEW

**Comment by mattip** [comment]
Any comments before I merge this DOC update?

**Comment by eric-wieser** [comment]
There's no easy way to preview this, right?

If all you've done is lift the wiki content without modifications, i reckon its good to go - it'll be easier to iterate on in future PRs.

File: doc/source/user/theory.broadcasting.rst [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+===========================
+Array Broadcasting in Numpy
+===========================
+
+.. note::
+   Originally part of the scipy.org wiki, available `here
+   <https://scipy.github.io/old-wiki/pages/EricsBroadcastingDoc>`_ or from the
+   `github repo
+   <https://github.com/scipy/old-wiki/blob/gh-pages/pages/EricsBroadcastingDoc.html>`_
```
Comment (rgommers) [inline doc/source/user/theory.broadcasting.rst]: this note is unnecessary (just noise for users), I suggest to remove it

File: doc/source/user/theory.broadcasting.rst [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+[Let's] ... explore a more advanced concept in numpy called broadcasting. The
```
Comment (rgommers) [inline doc/source/user/theory.broadcasting.rst]: start this like a normal sentence, "Let's"

File: doc/source/user/theory.broadcasting.rst [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+memory that slows computation. This article provides a gentle introduction to
+broadcasting with numerous examples ranging from simple to involved. It also
+provides hints on when and when not to use broadcasting.
+
+numpy operations are usually done element-by-element which requires two arrays
+to have exactly the same shape:
+
+.. code-block:: python
+    :caption: Example 1
+    :name: example-1
```
Comment (rgommers) [inline doc/source/user/theory.broadcasting.rst]: all of these `code-block` and captions can simply be dropped.

**Review by rgommers — CHANGES_REQUESTED** [review]
Overall this looks good, thanks Matti. Haven't checked the rendered version, but only a couple of minor comments on the diff. 

There's further cleanups that could be make, like `array` --> `np.array` in examples, but those are not essential.

File: doc/source/user/theory.broadcasting.rst (hunk 2 above)
Comment (mattip) [reply doc/source/user/theory.broadcasting.rst]: ok

File: doc/source/user/theory.broadcasting.rst (hunk 3 above)
Comment (mattip) [reply doc/source/user/theory.broadcasting.rst]: they provide an anchor for referencing them in the document, and provide nice formatting for copy-paste

File: doc/source/user/theory.broadcasting.rst (hunk 1 above)
Comment (mattip) [reply doc/source/user/theory.broadcasting.rst]: done

File: doc/source/user/theory.broadcasting.rst (hunk 3 above)
Comment (rgommers) [reply doc/source/user/theory.broadcasting.rst]: > they provide an anchor 

good point. okay, let's keep it as is.

> provide nice formatting for copy-paste

should be the same formatting as for a normal indented docstring example

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
Merged, thanks @mattip!

**Pull request merged**